Explain more MyST concepts, developer docs, and export docs

choldgraf opened the pull request:
This follows a little listening and learning session that we had at the Scientific Python 2025 workshop. Here are the major changes:

- It expands on our MyST Overview documentation as well as our Developer concepts, with the goal of making it easier to understand the MyST-specific concepts that others need to know in order to contribute.
- It updates much of the Exports docs, so that it is up to date and we document a few missing pieces of information.

### Related links

- https://github.com/myst-templates/article-theme/pull/1
- https://github.com/myst-templates/book-theme/pull/6
- closes https://github.com/jupyter-book/mystmd/issues/2014

stefanv commented on docs/developer.md:
```diff
@@ -61,32 +59,45 @@ Walking through this last example, we see that invoking `myst start --headless`
 The [`myst-to-react` package](https://github.com/jupyter-book/myst-theme/tree/main/packages/myst-to-react) provides a `<MyST />` component which can render MyST AST into a React tree.
 A React [context](https://react.dev/reference/react/useContext), named `ThemeContext` (defined [here in the `myst-theme` repository](https://github.com/jupyter-book/myst-theme/blob/main/packages/providers/src/theme.tsx)), is used to push state deeply into the tree, without having to pass it via props.
 
+#### Where to find renderers, themes, and templates
```
Are these three terms defined before they're used?

choldgraf replied:
I'll define them more clearly in the "overview" section

stefanv commented on docs/developer.md:
```diff
@@ -61,32 +59,45 @@ Walking through this last example, we see that invoking `myst start --headless`
+- [`github.com/jupyter-book/myst-theme`](https://github.com/jupyter-book/myst-theme): The core React-based renderer and theme.
+  - A collection of [MyST Rendering Packages](https://github.com/jupyter-book/myst-theme/tree/main/packages) that define various React UI components for default MyST themes to use.
+  - The [MyST React Renderer](https://github.com/jupyter-book/myst-theme/tree/main/packages/myst-to-react) generates React components out of MyST AST for use by the default MyST Themes.
+  - The [default MyST Themes](https://github.com/jupyter-book/myst-theme/tree/main/themes) are special out-of-the-box themes that we maintain, and that use the React Renderer.
```
Special how? If only in that they are out of the box, then can skip this word.

stefanv commented on docs/documents-exports.md:
```diff
@@ -30,13 +31,31 @@ Below are supported export types and links to documentation for further reading:
-You can also explore the [MyST templating](xref:jtex) documentation for a deeper dive into defining templates.
+## Configuration options for exports
+
+There are a few options you can use to configure exports:
```
Would the reader know where to put these options?

stefanv commented on docs/documents-exports.md:
````diff
@@ -62,7 +81,39 @@ project:
+```yaml
+- format: typst
+  id: typstpdf
+  template: https://github.com/rowanc1/typst-book.git
````
Or local path

rowanc1 commented on docs/developer.md:
```diff
@@ -221,9 +350,26 @@ npm run dev
+
+https://api.mystmd.org/templates/typst
+
+:::{error} todo - Document how the API server is configured and where it is run.
```
It is here:
https://github.com/myst-templates/templates

Again, no config.

We have also discussed removing this, which would enable the CLI to be more able to run off-line as well as removing a complexity/dependency.

choldgraf replied:
ah this is super helpful, thanks for pointing that out. I will add it!

I know that it seems trivial, but to folks that don't have the knowledge of "where does this thing exist? who maintains it? how is it configured?", explicitly documenting "there is no configuration for this and the logic is over here" is extremely helpful to others.

choldgraf commented:
OK I think this one is ready to go. We've already had a few reviews while it was in a "draft" mode. The latest commit adds a bit of explanation about what themes and templates are, and addresses the comments above.

rowanc1 commented:
Looks good from my side. 🚀

choldgraf commented on docs/developer.md:
```diff
@@ -61,49 +60,184 @@ Walking through this last example, we see that invoking `myst start --headless`
+
+In the diagram above, we saw that `mystmd` produces websites by:
+
+- Parsing a set of documents to an AST
```
```suggestion
- Parsing a set of documents to an AST.
```

choldgraf commented:
This PR triggered an interesting discussion amongst the team that I wanted to quickly encode here! We asked @stefanv if he wanted to merge, and he wasn't sure whether it was allowed for him to merge a PR or not.

This made us realize that we don't have team guidelines defined for things like "when to merge a PR, when to intentionally wait, etc". I opened up this issue to encode that conversation and try to land on lightweight practices that we can follow!

- https://github.com/jupyter-book/team-compass/issues/19

In this specific case, we discussed a rationale like the following for this one:

This is a significant documentation change, but it has approval from another person with merge rights (@rowanc1). Because this isn't a super complex or impactful change, anybody with merge rights should be encouraged to merge it. @stefanv should feel comfortable merging.

Thanks for the helpful and quick conversation @agoose77 and @fwkoch !

choldgraf commented:
OK because @stefanv was a slowpoke I used the opportunity to add a couple more things :-D 

Here's a little example of making a contribution using @fwkoch and @kne42's "edit button" example: https://github.com/jupyter-book/mystmd/pull/2012/commits/044e74f3d335ceaad06de10d288fa5d0d6b2222b

rowanc1 commented on docs/developer.md:
```diff
@@ -114,6 +114,22 @@ MyST has multiple renders, themes, and templates that allow it to transform MyST
 - explain styling
 :::
 
+#### Example: Adding an "edit this page" button
```
Love how easy this was. :)

stefanv commented on docs/overview.md:
```diff
@@ -1,26 +1,116 @@
+- First parse the Markdown `[some text](#a-label)`.
+- The result is a MyST AST node for a **Markdown link**.
+- Next, search the document AST for any Markdown link nodes with a target that starts with `#`. Check if there's [a target](./cross-references.md) for the link. If so, it is meant to be a cross reference.
+- If so, run a **Transformer** that converts the Markdown Link node into a Cross Reference node.
```
Just to make sure I understand: the rendered runs the transformer? Or there's a Big Transformer that runs the Little Transformers?

stefanv commented on docs/overview.md:
```diff
@@ -1,26 +1,116 @@
+- Check that figures have alt-texts
+- Convert non-standard AST nodes (e.g., ones generated from a custom user directive) into ones that MyST knows how to render[^ex-transform-node].
+
+[^ex-transform-node]: This is a pattern used in e.g. https://github.com/projectpythia-mystmd/cookbook-gallery/blob/main/pythia-gallery.py where an `executable transform` (non-JS transform that communicates over `STDIO` with `JSON`) takes custom `pythia-cookbooks` nodes and converts them (via some HTTP fetches) to a grid of cards by outputting the relevant grid and card AST nodes.
```
How do you communicate with JSON? Or is it communicating *in* JSON with .... the Python script?

stefanv commented on docs/overview.md:
````diff
@@ -1,26 +1,116 @@
+MyST {term}`themes` and MyST {term}`templates` are both related to _exporting_ a resolved MyST AST into one or more final output formats. They differ in the amount of _customizability_ that a user has in altering their behavior, but the end-result is the same (generating an export artifact).
+
+```{glossary}
+Templates
````
This definition of templates is clearer to me than the one above.

stefanv commented on docs/overview.md:
```diff
@@ -1,26 +1,116 @@
+: Templates are usually used for [static exports](./documents-exports.md) (like LaTeX or Typst). The are structured like a final document (e.g., a LaTeX template for a specific journal), but with fields that MyST uses to insert document metadata at export time (e.g., a `doc.author` field if you want to insert the author name). See [the `myst-templates` GitHub repository](https://github.com/myst-t ...
+
+Themes
+: Themes are a special-case of templates because they expose more customizability than a typical template does. They are usually meant for _websites_, because there is a lot more control a website exposes compared to a static document like LaTeX. Themes have more complex logic, and thus their source-code is often in a dedicated repository. For examples, see the [build outputs of the `book` and `a ...
```
Are themes special cases of templates, or are they just templates? And if they're just templates, why do we have a special name for them?

Sounds like "theme == web template".

stefanv reviewed (approved):
Sorry it took a while, @choldgraf, but I finished a read-through now. I caught a few small typos, and left a couple of higher level questions for future consideration. I think this is a solid improvement on the existing state, so I'd merge whenever you're ready.

choldgraf commented:
OK @stefanv I think that I answered all of your comments above. I had time on the flight so I made a few other improvements:

- I documented the release process for the myst-theme and mystmd, based on the one that we followed today.
- I explained a bit more how Executable Plugins work (cc @agoose77 )
- I cleaned up some of the theme and templating information to reduce redundant info
- I added more documentation about the "Edit this page" button
- I cleaned up some of the frontmatter docs as a result

I don't think there's anything too complex in there, mostly it's just making things a bit clearer and adding more explanation. IMO this is still safe to merge, but I'd be happy for somebody to give another quick look, or just go for it. Either way I'll merge tomorrow unless somebody asks me to hold off.

rowanc1 commented on docs/website-navigation.md:
```diff
@@ -140,6 +140,21 @@ site:
+
+To override this behavior and set a manual edit URL, use the `edit_url` field in [MyST frontmatter](#table-frontmatter).
+
+To disable the `Edit this page` button, set the value of `edit_url` to `none`.
```
Should this be `null` not `none`?

choldgraf replied:
I was pulling the text from the `edit_url` frontmatter docs here: https://mystmd.org/guide/frontmatter. But looking at the code I think that `null` is better so I'll switch it to that and @fwkoch should object if he thinks that's wrong!
